File: backend/app/services/ollama_service.py
import json
from collections.abc import AsyncGenerator

import httpx

from app.core.config import settings


class OllamaService:
    def __init__(self, base_url: str | None = None):
        self.base_url = base_url or settings.ollama_base_url

    async def list_models(self) -> list[dict]:
        async with httpx.AsyncClient(base_url=self.base_url, timeout=10.0) as client:
            resp = await client.get("/api/tags")
            resp.raise_for_status()
            data = resp.json()
            return data.get("models", [])

    async def stream_chat(
        self, model: str, messages: list[dict]
    ) -> AsyncGenerator[str, None]:
        """Stream assistant tokens from Ollama's /api/chat endpoint."""
        payload = {"model": model, "messages": messages, "stream": True}
        async with httpx.AsyncClient(base_url=self.base_url, timeout=None) as client:
            async with client.stream("POST", "/api/chat", json=payload) as resp:
                resp.raise_for_status()
                async for line in resp.aiter_lines():
                    if not line.strip():
                        continue
                    chunk = json.loads(line)
                    if chunk.get("done"):
                        break
                    token = chunk.get("message", {}).get("content", "")
                    if token:
                        yield token


ollama_service = OllamaService()

File: app/Services/OllamaService.php
<?php

namespace App\Services;

use Generator;
use Illuminate\Support\Facades\Http;

class OllamaService
{
    protected string $baseUrl;

    public function __construct()
    {
        $this->baseUrl = config('services.ollama.base_url', 'http://localhost:11434');
    }

    public function listModels(): array
    {
        $response = Http::get("{$this->baseUrl}/api/tags");
        $response->throw();

        return $response->json('models', []);
    }

    /**
     * Stream tokens from Ollama's /api/chat endpoint as they're generated.
     *
     * @param  array<int, array{role: string, content: string}>  $messages
     * @return Generator<string>
     */
    public function streamChat(string $model, array $messages): Generator
    {
        $response = Http::withOptions(['stream' => true])
            ->timeout(0)
            ->post("{$this->baseUrl}/api/chat", [
                'model' => $model,
                'messages' => $messages,
                'stream' => true,
            ]);

        $body = $response->toPsrResponse()->getBody();
        $buffer = '';

        while (! $body->eof()) {
            $buffer .= $body->read(1024);

            while (($pos = strpos($buffer, "\n")) !== false) {
                $line = trim(substr($buffer, 0, $pos));
                $buffer = substr($buffer, $pos + 1);

                if ($line === '') {
                    continue;
                }

                $chunk = json_decode($line, true);

                if (! $chunk) {
                    continue;
                }

                if (! empty($chunk['done'])) {
                    return;
                }

                $token = $chunk['message']['content'] ?? '';

                if ($token !== '') {
                    yield $token;
                }
            }
        }
    }
}
